File: tests-cypress/cypress/support/pages/dashboard/dashboard-page.ts
import { routes } from '../../constants/routes'
import { testId } from '../../utils/core/selectors'

export const dashboardObj = {
  page: testId('dashboard-page'),
  error: testId('dashboard-error'),
  stats: {
    portfolioValue: testId('stat-portfolio-value'),
    change24h: testId('stat-24h-change'),
    watchlist: testId('stat-watchlist'),
    watchlistValue: testId('stat-watchlist-value'),
    transactions: testId('stat-transactions'),
    transactionsValue: testId('stat-transactions-value'),
  },
  chartCard: testId('dashboard-chart-card'),
  gainers: testId('top-gainers'),
  losers: testId('top-losers'),
}

export const dashboardPage = {
  visit: () => {
    cy.visit(routes.dashboard)
    cy.get(dashboardObj.page).should('be.visible')
  },

  verifyStatsVisible: () => {
    cy.get(dashboardObj.stats.portfolioValue).should('be.visible')
    cy.get(dashboardObj.stats.change24h).should('be.visible')
    cy.get(dashboardObj.stats.watchlist).should('be.visible')
    cy.get(dashboardObj.stats.transactions).should('be.visible')
  },

  verifyChartVisible: () => {
    cy.get(dashboardObj.chartCard).should('be.visible')
  },

  verifyTopMoversVisible: () => {
    cy.get(dashboardObj.gainers).should('be.visible')
    cy.get(dashboardObj.losers).should('be.visible')
  },

  verifyTransactionsCount: (count: number) => {
    cy.get(dashboardObj.stats.transactionsValue).should('have.text', String(count))
  },

  verifyWatchlistCount: (count: number) => {
    cy.get(dashboardObj.stats.watchlistValue).should('have.text', String(count))
  },
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/pages/DashboardPage.java
package com.coinpulse.selenium.pages;

import com.coinpulse.selenium.constants.Routes;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;

/**
 * Dashboard page object; selectors mirror the Cypress dashboard page object.
 */
public class DashboardPage extends BasePage {

    private static final By PAGE = byTestId("dashboard-page");
    private static final By STAT_PORTFOLIO_VALUE = byTestId("stat-portfolio-value");
    private static final By STAT_24H_CHANGE = byTestId("stat-24h-change");
    private static final By STAT_WATCHLIST = byTestId("stat-watchlist");
    private static final By STAT_WATCHLIST_VALUE = byTestId("stat-watchlist-value");
    private static final By STAT_TRANSACTIONS = byTestId("stat-transactions");
    private static final By STAT_TRANSACTIONS_VALUE = byTestId("stat-transactions-value");
    private static final By CHART_CARD = byTestId("dashboard-chart-card");
    private static final By TOP_GAINERS = byTestId("top-gainers");
    private static final By TOP_LOSERS = byTestId("top-losers");

    public DashboardPage(WebDriver driver) {
        super(driver);
    }

    public void open() {
        openPath(Routes.DASHBOARD);
        awaitVisible(PAGE);
    }

    public void awaitLoaded() {
        awaitVisible(PAGE);
    }

    public void verifyStatsVisible() {
        awaitVisible(STAT_PORTFOLIO_VALUE);
        awaitVisible(STAT_24H_CHANGE);
        awaitVisible(STAT_WATCHLIST);
        awaitVisible(STAT_TRANSACTIONS);
    }

    public void verifyChartVisible() {
        awaitVisible(CHART_CARD);
    }

    public void verifyTopMoversVisible() {
        awaitVisible(TOP_GAINERS);
        awaitVisible(TOP_LOSERS);
    }

    public void verifyTransactionsCount(int count) {
        wait.until(ExpectedConditions.textToBe(STAT_TRANSACTIONS_VALUE, String.valueOf(count)));
    }

    public void verifyWatchlistCount(int count) {
        wait.until(ExpectedConditions.textToBe(STAT_WATCHLIST_VALUE, String.valueOf(count)));
    }
}
